components/support/rc_crypto/src/pbkdf2.rs

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{digest, error::*, hmac};
+use byteorder::{BigEndian, ByteOrder};
+
+/// Extend passwords using pbkdf2, based on the following [rfc](https://www.ietf.org/rfc/rfc2898.txt)
+/// # Arguments
+///
+///  * `passphrase` - The password to stretch
+///  * `salt` - A salt to use in the generation process
+///  * `iterations` - The number of iterations the hashing algorithm will run on each section of the key
+///  * `key_len` - The desired length of the output, in bytes
+///  * `algorithm` - The digest HMAC algorithm to use
+///
+/// # Examples
+///
+/// ```
+/// use rc_crypto::{digest, pbkdf2};
+/// let stretched = pbkdf2::pbkdf2_generate(b"password", b"salt", 2, 32, &digest::Algorithm::SHA256).unwrap(); // Oh oh should handle the error!
+/// assert_eq!(hex::encode(stretched), "ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43");
+//
+///```
+///
+/// # Erros
+///
+/// Could possibly return an error if the HMAC algorithm fails
+pub fn pbkdf2_generate(
+    passphrase: &[u8],
+    salt: &[u8],
+    iterations: u32,
+    key_len: u32,
```

Comment:
API nit: I think it would be more idiomatic to let the caller provide an output buffer `out: &mut [u8]` and have this function will it, rather than allocate a new buffer. See e.g. the API exposed by `hkdf.rs` in this same directory.